spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/WhitelistedSerializationStrategy.java

```diff
@@ -0,0 +1,146 @@
+package org.springframework.security.oauth2.common.util;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.NotSerializableException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamClass;
+import java.util.Collections;
+import java.util.List;
+
+import org.springframework.util.ClassUtils;
+
+/**
+ * This is a {@link SerializationStrategy} which uses a whitelist of allowed classes for deserialization.
+ *
+ * @author Artem Smotrakov
+ * @since 2.4
+ */
+public class WhitelistedSerializationStrategy implements SerializationStrategy {
```

Comment:
> a user can extend this class to use it

Even if a user were to extend this class, there aren't any `protected` methods where a user can override the behaviour. They would only be able to override `serialize` and/or `deserialize` to change behaviour. Instead of extension, a user can implement a delegation-based approach to achieve the same.

> Do you have a reason why this class has to be final?

To prevent extension so we are not on the hook for support. However, given that 2.4.0 is the last minor version and this project will soon be deprecated, it's pretty much a moot point. I would be fine if we left this as-is.